src/cpp/core/r_util/RParser.cpp

```diff
@@ -0,0 +1,893 @@
+         return;                                                               \
+      }                                                                        \
+   } while (0)
+
+#define ENSURE_TYPE(__CURSOR__, __STATUS__, __TYPE__)                          \
+   do                                                                          \
+   {                                                                           \
+      if (!__CURSOR__.isType(__TYPE__))                                        \
+      {                                                                        \
+         DEBUG("(" << __LINE__ << "): Expected "                               \
+                   << string_utils::wideToUtf8(typeToWideString(__TYPE__)));   \
+         __STATUS__.lint().unexpectedToken(                                    \
+             __CURSOR__, L"'" + typeToWideString(__TYPE__) + L"'");            \
+         return;                                                               \
+      }                                                                        \
+   } while (0)
+
+#define UNEXPECTED_TOKEN(__CURSOR__, __STATUS__)                               \
+   do                                                                          \
+   {                                                                           \
+      DEBUG("(" << __LINE__ << "): Unexpected token (" << __CURSOR__ << ")");  \
+      __STATUS__.lint().unexpectedToken(__CURSOR__);                           \
+   } while (0)
+
+void handleIdentifier(TokenCursor& cursor,
+                      ParseStatus& status)
+{
+   // Check to see if we are defining a symbol at this location.
+   // Note that both string and id are valid for assignments.
+   
+   // Don't cache identifiers if the previous or next tokens
+   // are 'extraction' operators (e.g. '$').
+   //
+   // TODO: Handle namespaced symbols explicitly.
+   if (isExtractionOperator(cursor.previousSignificantToken()) ||
+       isExtractionOperator(cursor.nextSignificantToken()))
+      return;
+   
+   if (cursor.isType(RToken::ID) ||
+       cursor.isType(RToken::STRING))
+   {
+      if (isLocalLeftAssign(cursor.nextSignificantToken()) ||
+          isLocalRightAssign(cursor.previousSignificantToken()))
+         status.node()->addDefinedSymbol(cursor);
+   }
+   
+   // If this is truly an identifier, add a reference.
+   if (cursor.isType(RToken::ID))
+      status.node()->addReferencedSymbol(cursor);
+}
+
+} // anonymous namespace
+
+void doParse(TokenCursor&, ParseStatus&);
+
+ParseResults parse(const std::string& rCode,
+                   const ParseOptions& parseOptions)
+{
+   return parse(string_utils::utf8ToWide(rCode), parseOptions);
+}
+
+ParseResults parse(const std::wstring& rCode,
+                   const ParseOptions& parseOptions)
+{
+   using namespace boost::timer;
+   
+   if (rCode.empty() ||
+       rCode.find_first_not_of(L" \n\t\v") == std::string::npos)
```

Comment:
Good point, thanks! @jjallaire -- can we safely assume that any user documents we encounter will either use Windows newlines (`\r\n`) or Linux newlines (`\n`), or should we operate under the assumption that there are some machines out there that might use just `\r` or even `\n\r`?